base/inference.jl

```diff
@@ -2927,36 +2927,43 @@ function optimize(me::InferenceState)
 
     if isa(me.bestguess, Const) || isconstType(me.bestguess)
         me.const_ret = true
-        ispure = me.src.pure
-        if !ispure && length(me.src.code) < 10
-            ispure = true
+        proven_pure = false
+        # must be proven pure to use const_api; otherwise we might skip throwing errors
+        # (issue #20704)
+        if length(me.src.code) < 10
+            proven_pure = true
             for stmt in me.src.code
                 if !statement_effect_free(stmt, me.src, me.mod)
-                    ispure = false
+                    proven_pure = false
                     break
                 end
             end
-            if ispure
+            if proven_pure
                 for fl in me.src.slotflags
                     if (fl & Slot_UsedUndef) != 0
-                        ispure = false
+                        proven_pure = false
                         break
                     end
                 end
             end
         end
-        me.src.pure = ispure
+        if proven_pure
+            me.src.pure = true
```

Comment:
This doesn't seem quite right, but do we ever actually look at this field? I suspect we should acknowledge that `source.pure` and `inferred.pure` and reflecting slightly different attributes (the former meaning `@pure`, the latter meaning `proven_pure`). There's a bit of a mis-match in terminology here though too, since we're using `me.src.pure` to mean the same as `const_api / jlcall_api == 2`. However, it's perfectly reasonable to know that a function is effect-free / pure, but not know what it returns.